zebra-network/src/peer_set/set.rs

```diff
@@ -390,38 +410,42 @@ where
         }
     }
 
-    /// Takes a ready service by key, preserving `preselected_p2c_index` if possible.
-    fn take_ready_service(&mut self, key: &D::Key) -> Option<(D::Key, D::Service)> {
-        if let Some((i, key, svc)) = self.ready_services.swap_remove_full(key) {
-            // swap_remove perturbs the position of the last element of
-            // ready_services, so we may have invalidated self.next_idx, in
-            // which case we need to fix it. Specifically, swap_remove swaps the
-            // position of the removee and the last element, then drops the
-            // removee from the end, so we compare the active and removed indices:
-            //
-            // We just removed one element, so this was the index of the last element.
-            let last_idx = self.ready_services.len();
-            self.preselected_p2c_index = match self.preselected_p2c_index {
-                None => None,                        // No active index
-                Some(j) if j == i => None,           // We removed j
-                Some(j) if j == last_idx => Some(i), // We swapped i and j
-                Some(j) => Some(j),                  // We swapped an unrelated service.
-            };
-            // No Heisenservices: they must be ready or unready.
-            assert!(!self.cancel_handles.contains_key(&key));
-            Some((key, svc))
+    /// Takes a ready service by key, invalidating `preselected_p2c_peer` if needed.
+    fn take_ready_service(&mut self, key: &D::Key) -> Option<D::Service> {
+        if let Some(svc) = self.ready_services.remove(key) {
+            if Some(*key) == self.preselected_p2c_peer {
+                self.preselected_p2c_peer = None;
+            }
+
+            assert!(
+                !self.cancel_handles.contains_key(key),
+                "cancel handles are only used for unready service work"
+            );
+
+            Some(svc)
         } else {
             None
         }
     }
 
+    /// Remove the service corresponding to `key` from the peer set.
+    ///
+    /// Drops the service, cancelling any pending request or response to that peer.
+    /// If the peer does not exist, does nothing.
     fn remove(&mut self, key: &D::Key) {
-        if self.take_ready_service(key).is_some() {
+        if let Some(ready_service) = self.take_ready_service(key) {
+            // A ready service has no work to cancel, so just drop it.
+            std::mem::drop(ready_service);
         } else if let Some(handle) = self.cancel_handles.remove(key) {
+            // Cancel the work, implicitly dropping the cancel handle.
+            // The service future returns a `Canceled` error,
+            // making `poll_unready` drop the service.
             let _ = handle.send(CancelClientWork);
         }
     }
 
+    /// Adds a busy service to the unready list.
+    /// Also adds a cancel handle for the service's current request.
```

Comment:
Nit: Maybe have the second sentence in the separate details section?

```suggestion
    /// Adds a busy service to the unready list.
    ///
    /// Also adds a cancel handle for the service's current request.
```

Reply:
The second sentence is essential to the operation of the method.

I tweaked the wording to make that clearer in ddc8fb036.